CustomCollections/CustomCollections/CustomHashTable.cs

```diff
@@ -0,0 +1,220 @@
+            Add(new KeyValuePair<TKey, TValue>(key, value));
+        }
+
+        public bool Remove(KeyValuePair<TKey, TValue> item)
+        {
+            return Remove(item.Key);
+        }
+
+        public bool Remove(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("key");
+            }
+            var position = GetPosition(key);
+            if (_hashTable[position].Key != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(_hashTable[position], new KeyValuePair<TKey, TValue>()))
+            {
+                _hashTable[position] = new KeyValuePair<TKey, TValue>();
+                Count--;
+                return true;
+            }
+            return false;
+        }
+
+        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        {
+            foreach (var item in _hashTable)
+            {
+                if (!item.Equals(null) || !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(item, new KeyValuePair<TKey, TValue>()))
+                {
+                    yield return item;
+                }
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private int GetPosition(TKey key)
+        {
+            var hash = key.GetHashCode();
+            var capacity = Capacity;
+            if (hash >= Capacity) capacity = hash + 1;
+            var pos = Math.Abs(hash % capacity);
+            return pos;
+        }
+
+        private ICollection<TKey> GetKeys()
+        {
+            List<TKey> keys = new List<TKey>();
+            foreach (var item in _hashTable)
+            {
+                if (item.Key != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(item, new KeyValuePair<TKey, TValue>()))
+                {
+                    keys.Add(item.Key);
+                }
+            }
+            return keys;
+        }
+
+        private ICollection<TValue> GetValues()
+        {
+            List<TValue> values = new List<TValue>();
+            foreach (var item in _hashTable)
+            {
+                if (item.Value != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(item, new KeyValuePair<TKey, TValue>()))
+                {
+                    values.Add(item.Value);
+                }
+            }
+            return values;
+        }
+
+        private bool HasAvailableMemory()
+        {
+            return Count + 1 < Capacity;
+        }
+
+        private void IncreaseSize()
+        {
+            var newHashTable = new KeyValuePair<TKey, TValue>[Capacity * 2];
+            for(int i = 0; i < Capacity; i++)
+            {
+                if (_hashTable[i].Key != null && !EqualityComparer<KeyValuePair<TKey, TValue>>.Default.Equals(_hashTable[i], new KeyValuePair<TKey, TValue>()))
+                {
+                    newHashTable[i] = _hashTable[i];
```

Comment:
Next time you search for item you'll receive different arrayindex, since your capacity changed and wouldn't be ever able to find it.